lmdeploy/pytorch/spec_decode/reject_sampler.py

```diff
@@ -0,0 +1,126 @@
+@record_function('rejection_sample')
+def rejection_sample(
+    target_probs: Tensor,
+    draft_token_ids: LongTensor,
+    bonus_token_ids: LongTensor,
+    sample_policy: SamplePolicy = SamplePolicy.ALL_GREEDY,
+    draft_probs: Optional[Tensor] = None,
+):
+    """rejection sample
+    Args:
+        target_probs (Tensor):
+
+    """
+    assert draft_probs is None or draft_probs.is_contiguous()
+    assert sample_policy == SamplePolicy.ALL_GREEDY, 'only support all greedy sampling policy'
+
+    target_argmax_tokens = target_probs.argmax(dim=-1)
+    return greedy_reject_sampler(draft_token_ids, target_argmax_tokens, bonus_token_ids)
+
+
+def greedy_reject_sampler(draft_token_ids, target_token_ids, bonus_token_ids):
+    """Greedy reject sampler
+    1. keep targets tokens that are equal to draft tokens
+    2. keep first not equal target tokens
+    3. add bonus tokens if all equal
+    Args:
+        draft_token_ids: (batch_size, num_spec_tokens)
+        target_token_ids: (batch_size, num_spec_tokens)
+        bonus_token_ids: (batch_size, 1)
+    Returns:
+        output_token_ids: (batch_size, num_spec_tokens + 1)
+    """
+    masks = draft_token_ids == target_token_ids
+    batch_size, num_spec_tokens = draft_token_ids.shape
+    # check rest draft tokens
+    range_data = torch.arange(num_spec_tokens, device=draft_token_ids.device)[None, :]
+    equals = (masks.cumsum(dim=1) - 1) == range_data
+    num_rejected_tokens = num_spec_tokens - equals.sum(dim=1)
+    first_diff_indices = torch.argmin(equals.int(), dim=1, keepdim=True)
+    keeps = range_data.repeat(batch_size, 1) <= first_diff_indices
+    keeps = keeps | equals
+    keep_token_ids = torch.where(keeps, target_token_ids, -1)
+    # add bonus tokens
+    keep_bonus_ids = torch.where(equals[:, -1:], bonus_token_ids, -1)
+    output_token_ids = torch.cat([keep_token_ids, keep_bonus_ids], dim=1)
+    # get last token ids
+    last_indices = (torch.cat([keeps, equals[:, -1:]], dim=1).cumsum(dim=1) - 1)[:, -1].flatten()
+    last_token_ids = output_token_ids[torch.arange(batch_size, device=draft_token_ids.device), last_indices]
+    return output_token_ids, num_rejected_tokens, last_token_ids
+
+
+if __name__ == '__main__':
+    num_seq = 3
+    max_spec_len = 2
+
+    num_draft_tokens = torch.LongTensor([2, 0, 1]).cuda()
+    vocab_size = 3
+    target_probs = torch.tensor([[0.5, 0.3, 0.2], [0.3, 0.5, 0.2], [0.1, 0.7, 0.2]], dtype=torch.bfloat16).cuda()
+    draft_token_ids = torch.tensor([0, 1, 2], dtype=torch.long).cuda()
+    bonus_token_ids = torch.tensor([2, 0, 2], dtype=torch.long).cuda()
+    expected_last_token_ids = torch.tensor([2, 0, 1], dtype=torch.long).cuda()
+    expected_num_rejected_tokens = torch.tensor([0, 0, 1], dtype=torch.long).cuda()
+    expected_output_token_ids = torch.tensor([[0, 1, 2], [0, -1, -1], [1, -1, -1]], dtype=torch.long).cuda()
+
+    draft_probs = None
+    target_probs, draft_token_ids, bonus_token_ids, max_spec_len = torch.load('tmp.pt')
```

Comment:
This test requires `tmp.pt`, do not place it here.
Could we add it in unit test?